datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs

```diff
@@ -0,0 +1,197 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use log::info;
+use sqllogictest::{ColumnType, DBOutput};
+use testcontainers::core::WaitFor;
+use testcontainers::images::generic::GenericImage;
+use tokio::task::JoinHandle;
+
+pub struct Postgres {
+    client: Arc<tokio_postgres::Client>,
+    join_handle: JoinHandle<()>,
+}
+
+pub const PG_USER: &str = "postgres";
+pub const PG_PASSWORD: &str = "postgres";
+pub const PG_DB: &str = "test";
+pub const PG_PORT: u16 = 5432;
+
+impl Postgres {
+    pub fn postgres_docker_image() -> GenericImage {
+        let postgres_test_data = match datafusion::test_util::get_data_dir(
+            "POSTGRES_TEST_DATA",
+            "tests/sqllogictests/postgres",
+        ) {
+            Ok(pb) => pb.display().to_string(),
+            Err(err) => panic!("failed to get arrow data dir: {err}"),
+        };
+        GenericImage::new("postgres", "15")
+            .with_wait_for(WaitFor::message_on_stderr(
+                "database system is ready to accept connections",
+            ))
+            .with_env_var("POSTGRES_DB", PG_DB)
+            .with_env_var("POSTGRES_USER", PG_USER)
+            .with_env_var("POSTGRES_PASSWORD", PG_PASSWORD)
+            .with_env_var(
+                "POSTGRES_INITDB_ARGS",
+                "--encoding=UTF-8 --lc-collate=C --lc-ctype=C",
+            )
+            .with_exposed_port(PG_PORT)
+            .with_volume(
+                format!(
+                    "{0}/csv/aggregate_test_100.csv",
+                    datafusion::test_util::arrow_test_data()
+                ),
+                "/opt/data/csv/aggregate_test_100.csv",
+            )
+            .with_volume(
+                format!("{0}/postgres_create_table.sql", postgres_test_data),
+                "/docker-entrypoint-initdb.d/0_create_table.sql",
+            )
+    }
+
+    pub async fn connect_with_retry(
```

Comment:
In my implementation a docker container starts before each test. And sometimes first connections fail. I didn't find a solution to deterministically define when a container is fully ready, so I introduced a connection retry.

Reply:
👍 

I think in general the approach of orchestrating the docker containers using github CI worked well rather than restarting the containers within the tests